contracts/pcv/PCVSwapperUniswap.sol

```diff
@@ -0,0 +1,199 @@
+        require(_maxSpentPerSwap != 0, "PCVSwapperUniswap: Cannot swap 0.");
+        maxSpentPerSwap = _maxSpentPerSwap;
+    }
+
+    /// @notice Sets the maximum amount of tokens to buy
+    /// @param _tokenBuyLimit the amount of tokens above which swaps will revert
+    function setTokenBuyLimit(uint256 _tokenBuyLimit) external onlyGovernor {
+      tokenBuyLimit = _tokenBuyLimit;
+    }
+
+    /// @notice sets the minimum time between swaps
+    function setSwapFrequency(uint256 _duration) external onlyGovernor {
+       _setDuration(_duration);
+    }
+
+    /// @notice Get the minimum time between swaps
+    /// @return the time between swaps
+    function getSwapFrequency() external view returns (uint256) {
+      return duration;
+    }
+
+    /// @notice Get the token to spend
+    /// @return The address of the token to spend
+    function getTokenSpent() external view override returns (address) {
+      return tokenSpent;
+    }
+
+    /// @notice Get the token to receive
+    /// @return The address of the token to receive
+    function getTokenReceived() external view override returns (address) {
+      return tokenReceived;
+    }
+
+    /// @notice Get the address receiving the inbound swapped tokens
+    /// @return The address receiving tokens
+    function getReceivingAddress() external view override returns (address) {
+      return tokenReceivingAddress;
+    }
+
+    /// @notice Swap tokenSpent for tokenReceived
+    function swap() external override afterTime whenNotPaused {
+      uint256 balance = IERC20(tokenSpent).balanceOf(address(this));
+      require(balance != 0, "PCVSwapperUniswap: no tokenSpent left.");
+      if (tokenBuyLimit != 0) {
+        require(IERC20(tokenReceived).balanceOf(tokenReceivingAddress) < tokenBuyLimit, "PCVSwapperUniswap: tokenBuyLimit reached.");
+      }
+
+      // Get pair reserves
+      (uint256 _token0, uint256 _token1) = getReserves();
+      (uint256 tokenSpentReserves, uint256 tokenReceivedReserves) =
+          pair.token0() == tokenSpent
+              ? (_token0, _token1)
+              : (_token1, _token0);
+
+      // Prepare swap
+      uint256 amount = Math.min(maxSpentPerSwap, balance);
+      uint256 amountOut = UniswapV2Library.getAmountOut(
+        amount,
+        tokenSpentReserves,
+        tokenReceivedReserves
+      );
+
+      // Check spot price vs oracle price discounted by max slippage
+      // E.g. for a max slippage of 3%, spot price must be >= 97% oraclePrice
+      (Decimal.D256 memory twap, bool oracleValid) = oracle.read();
+      require(oracleValid, "PCVSwapperUniswap: invalid oracle.");
+      uint256 oracleAmountOut = twap.mul(amount).asUint256();
+      Decimal.D256 memory maxSlippage = Decimal.ratio(BASIS_POINTS_GRANULARITY - maximumSlippageBasisPoints, BASIS_POINTS_GRANULARITY);
+      Decimal.D256 memory oraclePriceMinusSlippage = maxSlippage.mul(oracleAmountOut);
```

Comment:
there's a problem of decimals here if both tokens have different decimals (e.g. WETH has 18 & USDC has 6)

Reply:
fixed